dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs

```diff
@@ -0,0 +1,153 @@
+
+        // Import sample functions.
+        kernel.ImportPluginFromFunctions("HelperFunctions",
+        [
+            kernel.CreateFunctionFromMethod(() => DateTime.UtcNow.ToString("R"), "GetCurrentUtcTime", "Retrieves the current time in UTC."),
+            kernel.CreateFunctionFromMethod((string cityName) =>
+                cityName switch
+                {
+                    "Boston" => "61 and rainy",
+                    "London" => "55 and cloudy",
+                    "Miami" => "80 and sunny",
+                    "Paris" => "60 and rainy",
+                    "Tokyo" => "50 and sunny",
+                    "Sydney" => "75 and sunny",
+                    "Tel Aviv" => "80 and sunny",
+                    _ => "31 and snowing",
+                }, "GetWeatherForCity", "Gets the current weather for the specified city"),
+        ]);
+
+        // Enable automatic function calling.
+        var executionSettings = new OpenAIPromptExecutionSettings { ToolCallBehavior = ToolCallBehavior.AutoInvokeKernelFunctions };
+
+        // Invoke prompt with arguments.
+        const string Prompt = "Given the current time of day and weather, what is the likely color of the sky in {{$city}}?";
+        var result = await kernel.InvokePromptAsync(Prompt, new(executionSettings) { ["city"] = "Boston" });
+
+        Console.WriteLine(result);
+    }
+
+    /// <summary>
+    /// Filter which logs an information available during function invocation such as:
+    /// Function name, arguments, execution settings, result, duration, token usage.
+    /// </summary>
+    private sealed class FunctionInvocationLoggingFilter(ILogger logger) : IFunctionInvocationFilter
+    {
+        public async Task OnFunctionInvocationAsync(FunctionInvocationContext context, Func<FunctionInvocationContext, Task> next)
+        {
+            long startingTimestamp = Stopwatch.GetTimestamp();
+
+            logger.LogInformation("Function {FunctionName} invoking.", context.Function.Name);
+
+            if (context.Arguments.Count > 0)
+            {
+                logger.LogTrace("Function arguments: {Arguments}", context.Arguments);
+            }
+
+            if (logger.IsEnabled(LogLevel.Information) && context.Arguments.ExecutionSettings is not null)
+            {
+                logger.LogInformation("Execution settings: {Settings}", JsonSerializer.Serialize(context.Arguments.ExecutionSettings));
+            }
+
+            await next(context);
+
+            TimeSpan duration = new((long)((Stopwatch.GetTimestamp() - startingTimestamp) * (10_000_000.0 / Stopwatch.Frequency)));
+
+            logger.LogInformation("Function {FunctionName} succeeded.", context.Function.Name);
+            logger.LogTrace("Function result: {Result}", context.Result.ToString());
+
+            if (logger.IsEnabled(LogLevel.Information))
+            {
+                logger.LogInformation("Function completed. Duration: {Duration}s", duration.TotalSeconds);
```

Comment:
I realise this is only an example, but milliseconds is probably more appropriate, since most function calls should not take seconds to complete, and it'll be easier to not work with fractions.

Reply:
All telemetry in this example is recreated from already existing telemetry in Semantic Kernel, based on requirement from original issue. Here is a log message from SK:
https://github.com/microsoft/semantic-kernel/blob/4ebd11b05d78e2f1800bbeae8b3c795c1d66a1ca/dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionLogMessages.cs#L116-L118